services/src/contexts/migrations/migration_0015_snapshot.sql

```diff
@@ -720,14 +783,400 @@ CREATE TYPE "DataProviderDefinition" AS (
+    "name" text,
+    no_data_value double precision,
+    data_type "RasterDataType"
+);
+
+CREATE TYPE "StacZone" AS (
+    "name" text,
+    epsg oid
+);
+
+CREATE TYPE "StacApiRetries" AS (
+    number_of_retries bigint,
+    initial_delay_ms bigint,
+    exponential_backoff_factor double precision
+);
+
+CREATE TYPE "GdalRetries" AS (
+    number_of_retries bigint
+);
+
+CREATE TYPE "StacQueryBuffer" AS (
+    start_seconds bigint,
+    end_seconds bigint
+);
+
+CREATE TYPE "SentinelS2L2ACogsProviderDefinition" AS (
+    "name" text,
+    id uuid,
+    api_url text,
+    bands "StacBand" [],
+    zones "StacZone" [],
+    stac_api_retries "StacApiRetries",
+    gdal_retries "GdalRetries",
+    cache_ttl int,
+    description text,
+    priority smallint,
+    query_buffer "StacQueryBuffer"
+);
+
+CREATE TYPE "CopernicusDataspaceDataProviderDefinition" AS (
+    "name" text,
+    id uuid,
+    stac_url text,
+    s3_url text,
+    s3_access_key text,
+    s3_secret_key text,
+    description text,
+    priority smallint,
+    gdal_config "StringPair" []
+);
+
+CREATE TYPE "ProDataProviderDefinition" AS (
+    -- one of
+    sentinel_s2_l2_a_cogs_provider_definition
+    "SentinelS2L2ACogsProviderDefinition",
+    copernicus_dataspace_provider_definition
+    "CopernicusDataspaceDataProviderDefinition"
+);
+
+CREATE TABLE pro_layer_providers (
+    id uuid PRIMARY KEY,
+    type_name text NOT NULL,
+    name text NOT NULL,
+    definition "ProDataProviderDefinition" NOT NULL,
+    priority smallint NOT NULL DEFAULT 0
+);
+
+-- TODO: distinguish between roles that are (correspond to) users
+--       and roles that are not
+
+-- TODO: integrity constraint for roles that correspond to users
+--       + DELETE CASCADE
+
+CREATE TABLE roles (
+    id uuid PRIMARY KEY,
+    name text UNIQUE NOT NULL
+);
+
+CREATE TABLE users (
+    id uuid PRIMARY KEY REFERENCES roles (id),
+    email character varying(256) UNIQUE,
+    password_hash character varying(256),
+    real_name character varying(256),
+    active boolean NOT NULL,
+    quota_available bigint NOT NULL DEFAULT 0,
+    quota_used bigint NOT NULL DEFAULT 0,
+    -- TODO: rename to total_quota_used?
+    CONSTRAINT users_anonymous_ck CHECK ((
+        email IS NULL
+        AND password_hash IS NULL
+        AND real_name IS NULL
+    )
+    OR (
+        email IS NOT NULL
+        AND password_hash IS NOT NULL
+        AND real_name IS NOT NULL
+    )
+    ),
+    CONSTRAINT users_quota_used_ck CHECK (quota_used >= 0)
+);
+
+-- relation between users and roles
+
+-- all users have a default role where role_id = user_id
+
+CREATE TABLE user_roles (
+    user_id uuid REFERENCES users (id) ON DELETE CASCADE NOT NULL,
+    role_id uuid REFERENCES roles (id) ON DELETE CASCADE NOT NULL,
+    PRIMARY KEY (user_id, role_id)
+);
+
+CREATE TABLE user_sessions (
+    user_id uuid REFERENCES users (id) ON DELETE CASCADE NOT NULL,
+    session_id uuid REFERENCES sessions (id) ON DELETE CASCADE NOT NULL,
```

Comment:
does the sessions table still make sense? all sessions are user_sessions now and I think we only made the distinction because there were no users in free.

Reply:
[27e3b49](https://github.com/geo-engine/geoengine/pull/1009/commits/27e3b4995d355dbfdb4b116de7f385fc37f7b0f7)